integration/networking/port_mapping_linux_test.go

```diff
@@ -400,3 +400,43 @@ func TestAccessPublishedPortFromRemoteHost(t *testing.T) {
 		}
 	}
 }
+
+func TestRestartUserlandProxyUnder2MSL(t *testing.T) {
+	skip.If(t, testEnv.IsRootless())
+
+	ctx := setupTest(t)
+
+	d := daemon.New(t)
+	d.StartWithBusybox(ctx, t)
+	defer d.Stop(t)
+
+	c := d.NewClientT(t)
+	defer c.Close()
+
+	const netName = "nat-time-wait"
+	network.CreateNoError(ctx, t, c, netName,
+		network.WithDriver("bridge"),
+		network.WithOption(bridge.BridgeName, netName))
+	defer network.RemoveNoError(ctx, t, c, netName)
+
+	ctrOpts := []func(*container.TestContainerConfig){
+		container.WithName(sanitizeCtrName(t.Name() + "-server")),
+		container.WithExposedPorts("80/tcp"),
+		container.WithPortMap(nat.PortMap{"80/tcp": {{HostPort: "1780"}}}),
+		container.WithCmd("httpd", "-f"),
+		container.WithNetworkMode(netName),
+	}
+
+	serverID := container.Run(ctx, t, c, ctrOpts...)
+	defer c.ContainerRemove(ctx, serverID, containertypes.RemoveOptions{Force: true})
+
+	httpClient := &http.Client{Timeout: 3 * time.Second}
+	resp, err := httpClient.Get("http://127.0.0.1:1780")
+	assert.NilError(t, err)
+	assert.Check(t, is.Equal(resp.StatusCode, 404))
```

Comment:
I was actually curious if this would pass if no container was created, but looks like that gives a `dial tcp 127.0.0.1:1780: connect: connection refused`

Regardless; wondering if we should have something that's more clearly "it works", because I had trouble understanding the test here, and thought `404` was meant to indicate "should not return a result".

I guess we could either add some file to serve, or use some file that's expected to be present (e.g.) `http://127.0.0.1:1780/etc/resolv.conf` (but maybe that's confusing).

Reply:
> I guess we could either add some file to serve, or use some file that's expected to be present (e.g.) http://127.0.0.1:1780/etc/resolv.conf (but maybe that's confusing).

There are other tests doing exactly that in this suite. Generally speaking, we don't care at all about what's happening in L7. It's just an easy way to realistically test L4 (full TCP connection lifecycle, exchange a message back-and-forth, etc...).

I don't think we need to make things more complicated, like serving a real file, etc... That wouldn't add any value. But I added an inline comment stating that we don't care about the 404 response itself, and that we're just looking for a full-duplex TCP connection that we know is working correctly.

Reply:
Yeah, I'm fine with keeping the 404 for now. Perhaps we could find some other minimal approach. My main train of thought here was that usually a `404` means "something went wrong!" which could be confusing (at least it initially confused me).

Reply:
nit: Actually wondering now if the response needs to be checked at all for this; i.e., the `err` returned will be non-nil when failing to connect.

Reply:
Right, we don't need to check the status code. I'll open a follow-up PR to fix the few tests that use this pattern.